### PR TITLE
2647: Fix next recurrence of events

### PR DESCRIPTION
**Pull request by steffenkleinle**
### Short description

Fix the displaying of the next recurrence of events if its near.

### Proposed changes

- Substract timezone offset difference from dates passed to rrule to account for usage of "local time" 
- Substract duration from dates passed to rrule to include currently happening events

### Side effects

None.

### Resolved issues

Fixes: #2647.

---

### CODE REVIEW

File: release-notes/unreleased/2647-next-recurrence.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+issue_key: 2467
+show_in_stores: true
+platforms:
+  - web
+  - ios
+  - android
+en: Fixed displaying of next date of an event if its near.
```
Comment (LeandraH) [inline release-notes/unreleased/2647-next-recurrence.yml:7]:
```suggestion
en: Fixed displaying of next date of an event if it's soon.
```

File: shared/api/models/__tests__/DateModel.spec.ts [hunk 2]
```diff
@@ -141,8 +141,8 @@ describe('DateModel', () => {
         recurrenceRule: rrulestr('DTSTART:20230414T050000\nRRULE:FREQ=WEEKLY;BYDAY=MO'),
       })
 
-      expect(date.startDate).toEqual(DateTime.fromISO('2023-10-09T07:00:00.000+02:00'))
-      expect(date.endDate).toEqual(DateTime.fromISO('2023-10-10T09:00:00.000+02:00'))
+      expect(date.startDate).toEqual(DateTime.fromISO('2023-10-02T07:00:00.000+02:00'))
```
Comment (f1sh1918) [inline shared/api/models/__tests__/DateModel.spec.ts:144]: why did you change this?

File: shared/api/models/DateModel.ts [hunk 3]
```diff
@@ -127,6 +128,17 @@ class DateModel {
...
+  private currentDateToRrule(): Date {
+    // rrule is not correctly handling timezones and is expecting dates in "local time", i.e. in the timezone of dstart
+    // Therefore we have to manually apply timezone offset, e.g. for daylight savings time
+    // https://github.com/jkbrzt/rrule#important-use-utc-dates
+    // Also include dates happening at the moment by subtracting the duration
+    return DateTime.now()
+      .minus(this._duration)
+      .minus({ minutes: this._offset - DateTime.now().offset })
+      .toJSDate()
+  }
```
Comment (LeandraH) [inline shared/api/models/DateModel.ts:140]: I am getting a little confused by the adding and subtracting of different offsets here and in the previous method. Maybe we can instead use `Luxon` methods like `setZone`?

File: shared/api/models/DateModel.ts [hunk 4]
```diff
@@ -127,6 +128,17 @@ class DateModel {
     return dateTime.plus({ minutes: this._offset - dateTime.offset })
   }
 
+  private currentDateToRrule(): Date {
+    // rrule is not correctly handling timezones and is expecting dates in "local time", i.e. in the timezone of dstart
+    // Therefore we have to manually apply timezone offset, e.g. for daylight savings time
+    // https://github.com/jkbrzt/rrule#important-use-utc-dates
+    // Also include dates happening at the moment by subtracting the duration
+    return DateTime.now()
+      .minus(this._duration)
```
Comment (LeandraH) [inline shared/api/models/DateModel.ts:137]: Why are we subtracting the duration here?

**Comment by f1sh1918** [comment]
@steffenkleinle can you describe how to reproduce this issue.
Just created a new event as described in issue description and can't reproduce it
https://webnext.integreat.app/testumgebung/de/events/test-recurrence

**Review by LeandraH — COMMENTED** [review]
I wasn't able to reproduce the original issue without mocking (did manage to lock myself out of a few accounts by changing time zones on my laptop though, yay 😅 ) but I follow your reasoning. But I would prefer methods with clearer names to not turn the `rruleToDateTime` and `currentDateToRrule` methods into ones that I for one would be scared to touch in the future.

**Review by f1sh1918 — APPROVED** [review]
looks fine to me

File: shared/api/models/DateModel.ts (hunk 4 above)
Comment (f1sh1918) [reply shared/api/models/DateModel.ts:137]: to show events that already started but didn't end yet

File: shared/api/models/DateModel.ts (hunk 3 above)
Comment (steffenkleinle) [reply shared/api/models/DateModel.ts:140]: I didn't figure out how to make that work, since (afaik) it only changes the timezone, the actual time (relative to UTC stays the same). I.e. it changes from 00:00 UTC to 01:00 UTC +1 which is exactly the same point in time, just in a different timezone.

**Comment by steffenkleinle** [comment]
> @steffenkleinle can you describe how to reproduce this issue. Just created a new event as described in issue description and can't reproduce it https://webnext.integreat.app/testumgebung/de/events/test-recurrence

Its only possible to reproduce by creating the event during summer time (or probably easier, mocking the rrule).

**Comment by steffenkleinle** [comment]
> I wasn't able to reproduce the original issue without mocking (did manage to lock myself out of a few accounts by changing time zones on my laptop though, yay 😅 ) but I follow your reasoning. But I would prefer methods with clearer names to not turn the `rruleToDateTime` and `currentDateToRrule` methods into ones that I for one would be scared to touch in the future.

@LeandraH what names do you propose? I am not sure if I get your point :see_no_evil:

File: shared/api/models/__tests__/DateModel.spec.ts (hunk 2 above)
Comment (steffenkleinle) [reply shared/api/models/__tests__/DateModel.spec.ts:144]: Now recurring events that are currently happening are also shown. Therefore the first recurrence is the 02.10.2023, not just the 09.10.2023 (as the mocked date is the 02.10.2023). However, I just changed the mocked date now.

File: shared/api/models/DateModel.ts (hunk 4 above)
Comment (LeandraH) [reply shared/api/models/DateModel.ts:137]: Ah, nice, thanks!

**Comment by LeandraH** [comment]
> > I wasn't able to reproduce the original issue without mocking (did manage to lock myself out of a few accounts by changing time zones on my laptop though, yay 😅 ) but I follow your reasoning. But I would prefer methods with clearer names to not turn the `rruleToDateTime` and `currentDateToRrule` methods into ones that I for one would be scared to touch in the future.
> 
> @LeandraH what names do you propose? I am not sure if I get your point 🙈

Sorry, poor phrasing, I didn't mean the names `rruleToDateTime` and `currentDateToRrule`, I meant the adding and subtracting in there, like I also wrote on those lines, and you already answered. I think the names `rruleToDateTime` and `currentDateToRrule` are pretty clear 👍

**Review by LeandraH — APPROVED** [review]
